sympy/series/limits.py

```diff
@@ -261,7 +262,8 @@ def set_signs(expr):
                         return expr.args[0] if abs_flag else S.One
             return expr
 
-        if e.has(Float):
+        inexact = e.has(Float)
+        if inexact:
             # Convert floats like 0.5 to exact SymPy numbers like S.Half, to
             # prevent rounding errors which can lead to unexpected execution
             # of conditional blocks that work on comparisons
```

Comment:
Maybe the proper fix is to identify where these comparisons are and fix them to work with floats.

Reply:
From #19453
> The cause of this is a rounding error. In this block of code
> 
> https://github.com/sympy/sympy/blob/b4777fdcef467b7132c055f8ac2c9a5059e6a145/sympy/core/mul.py#L1910-L1915
> 
> `sum(powers)` returns the sum of `(-0.333333333333333, 0.333333333333333)` as `-5.55111512312578e-17`.
> An alternative to get the correct answer is to use exact SymPy numbers, i.e. `S.Half` instead of `0.5`.
> 
> ```python
> In [1]: beta = Symbol("beta", real=True, positive=True)
>    ...: h = Symbol("h", real=True, positive=True)
>    ...: m = Symbol("m", real=True, positive=True)
>    ...: w = Symbol("omega", real=True, positive=True)
>    ...: g = Symbol("g", real=True, positive=True)
>    ...: 
>    ...: e = exp(1)
>    ...: q = 3*h**2*beta*g*e**(S.Half*h*beta*w)
>    ...: p = m**2*w**2
>    ...: s = e**(h*beta*w) - 1
>    ...: Z = -q/(4*p*s) - q/(2*p*s**2) - q*(e**(h*beta*w)+1)/(2*p*s**3) + e**(S.Half*h*beta*w)/s
>    ...: 
>    ...: E = -diff(log(Z), beta)
>    ...: limit(E - S.Half*h*w, beta, oo), limit(E.simplify() - S.Half*h*w, beta, oo)
> Out[1]: (0, 0)
> ```

I can't think of methods to change this comparison and make it work with floats without hardcoding a tolerance value. It also doesn't appear that much can be changed before the comparison (I doubt there is an alternative to `sum` that would return 0, as that is not expected from Python).

Do you have any suggestions?

Reply:
> we shouldn't apply nsimplify to the whole expression but rather target the relevant parts (the exponents).

This would fix the exponents making test_issue_19453 pass:
```
--- a/sympy/core/mul.py
+++ b/sympy/core/mul.py
@@ -1964,6 +1964,8 @@ def coeff_exp(term, x):
             ords3 = [coeff_exp(term, x) for term in fac]
             coeffs, powers = zip(*ords3)
             power = sum(powers)
+            if abs(power) < sum(abs(_) for _ in powers)/10**15:
+                power = S.Zero
             if (power - n).is_negative:
                 res += Mul(*coeffs)*(x**power)
```
However, the test will take 54.8 s vs. 3.1 s on the current current master. The reason for the slowdown seems to be the same as in this comment: https://github.com/sympy/sympy/blob/6ea48db2bd108a0c5b3c53aa2f1ef9e5d617bf3c/sympy/series/gruntz.py#L667-L671 
It seems that the series expansions are currently working but slow.

It also looks like applying nsimplify to exponents only may not suffice as gruntz will create new exponents here:  https://github.com/sympy/sympy/blob/6ea48db2bd108a0c5b3c53aa2f1ef9e5d617bf3c/sympy/series/gruntz.py#L637-L646
(c will usually be a Float if there are other Floats.)

So it seems to me that applying nsimplify to the whole expression would be the natural choice at least in gruntz.